### DISCUSSION TITLE
Realtime and RLS issue

### DISCUSSION THREAD

**Discussion by LuisErnestoZamb** [supabase/supabase#26980] [Questions] [answered]
Hello, I am doing an app in react-native, I have coded a realtime chat and works, however, once I apply the RLS the chat stop. 

I notice that my RLS policies works under supabase dev interface, in other words, it is filtering according to what I want, but I cannot receive realtime data after the RLS is applied in the react-native side. I am not sure if this is related with https://github.com/orgs/supabase/discussions/22484 If this is the case, is there a way to go around this issue temporally modifying the RLS?

Here are my policies so far:
```
CREATE POLICY select_own_messages ON messages
FOR SELECT
USING (
    EXISTS (
        SELECT 1
        FROM match
        JOIN profile ON match.user_a = profile.id OR match.user_b = profile.id
        WHERE match.id = messages.match_id
          AND profile.user_id = auth.uid()
    )
);
```

```
CREATE POLICY create_messages ON messages
FOR INSERT
WITH CHECK (
    EXISTS (
        SELECT 1
        FROM match
        JOIN profile ON match.user_a = profile.id OR match.user_b = profile.id
        WHERE match.id = messages.match_id
          AND profile.user_id = auth.uid()
    )
);
```
See video:
[![Here a video](https://github.com/supabase/supabase/assets/9245605/25f93b87-fd34-4b63-ac68-45a3ef0c0a4e)](https://www.screencast.com/t/9lXNi6WY)

**Comment by GaryAustin1** [comment]
Your link is not involved with postgres changes.  
You are not clear on what "stops".  Do your tables work and just realtime does not?  
Realtime only uses the select RLS to decide if the user gets notice.

**Reply by LuisErnestoZamb** [reply]
Hi @GaryAustin1 that's why I have edited the initial comment to add the video. The "select" works in supabase dev-web, but not in the react side. I can record another video with audio if is needed. I have updated the supabase dev (docker) version and also the package to "@supabase/supabase-js": "^2.43.4", and still not working.

**Reply by GaryAustin1** [reply]
For some reason then your Supabase client when it connects to realtime does not have a user session, or your filters are not correct.

You don't show your subscription code.
You can also do .subscribe(status=>console.log(status)) to get more info on connection state.

You can use the realtime monitor in the Dashboard with impersonation to see if realtime is working with your RLS.  This would help point to your client code not have a session in realtime or not.

**Reply by LuisErnestoZamb** [reply]
```
useEffect(() => {
 const subscription = supabase
      .channel('public:messages')
      .on('postgres_changes',
        {
          event: 'INSERT',
          schema: 'public',
          table: 'messages',
          filter: `match_id=eq.${match_id}`,
        }, payload => {
          setMessages(prevMessages => [payload.new, ...prevMessages]);
        })
      .subscribe(status => console.log(status));

    return () => {
      supabase.removeChannel(subscription);
    };
  }, []);

```
This is the code @GaryAustin1 ; once the RLS is disabled the app start to receive the messages.
The console.log(status) show: SUBSCRIBED
I did the realtime monitor in the Dashboard with impersonation and It is NOT filtering anything in realtime, I see the whole messages, so the RLS is not working on here (it is the opposite behaviour that I have in the app). On the other hand, the table editor Dashboard is filtering with the RLS properly.

**Reply by GaryAustin1** [reply]
Did it work in the realtime monitor in the dashboard with impersonation (not clear in your response)? If so then for some reason your client session is not set to with a user session at the point you do the realtime call is my guess.  RLS working in the table editor and the realtime monitor with impersonation points to the client not having the user session.

**Reply by LuisErnestoZamb** [reply]
The RLS is not working on the "realtime monitor in the dashboard with impersonation" as it is not filtering the messages, indeed, I am watching all of them.
<img width="1538" alt="Screenshot 2024-06-03 at 11 35 47 PM" src="https://github.com/supabase/supabase/assets/9245605/dcae8506-e8cf-4106-819b-dde201c92a99">

**Reply by GaryAustin1** [reply]
Out for the night...
If you are saying the dashboard does not see the realtime changes then your select RLS is not being met by the user observing for changes.  You could have a user able to insert, but a different user selecting may not work.
In your video are you sending and receiving from the same user id?  If not then you need to make sure the user "receiving" can select the table.   You maybe doing this, but hard to tell from the video...
Good luck.

**Reply by LuisErnestoZamb** [reply]
I see all the messages on the dashboard, but once I do the impersonation, the RLS is not filtering according to what I want, in fact, it is showing all the messages. This is not happening in the App (which is not receiving anything). Regarding the "user id", I don't think is the problem, as the table editor dashboard is showing me the right result.

So, I wonder why is showing me the right results on the table editor dashboard and not for the realtime? Here another video, with the WHOLE process:

[![Here a video](https://github.com/supabase/supabase/assets/9245605/72560033-9fa6-49e9-869d-c8798ca2a8f1)](https://app.screencast.com/PYh8oEfeQO1wH)

**Reply by LuisErnestoZamb** [reply]
@GaryAustin1 do you think I should become this discussion to a issue?, is there something else that I should do before scale this?

**Reply by GaryAustin1** [reply]
There is not an issue with RLS and realtime in general. I doubt this a bug versus something in your code or RLS but up to you.

**Reply by LuisErnestoZamb** [reply]
It is not the code as this work once the RLS are disable. Also, I did an update on my policy to something simple:
```
CREATE POLICY select_own_messages ON messages
FOR SELECT
USING (
    auth.uid() = user_id
);
```
It is not applying the policy under Realtime Dashboard, but yes under Table Editor Dashboard. Why this discrepancy?

**Reply by LuisErnestoZamb** [reply]
Update: The RLS and the Realtime Dashboard works on PRODUCTION but not in supabase CLI (local dev), don't even by using supabase-beta works, I have't tried my app directly on production but everything point to that Realtime don't work properly on dev.

**Comment by NekodRider** [comment, marked as answer]
I encountered the same issue and am able to fix it by setting `detectSessionInUrl:false`.

With `detectSessionInUrl:true`, the realtime channel can only receive `delete` events like `anon` role could.
